src/availability/data_source.rs

```diff
@@ -35,27 +35,56 @@ use std::{cmp::Ordering, error::Error, fmt::Debug, ops::RangeBounds};
     Ord(bound = ""),
     Hash(bound = "")
 )]
-pub enum ResourceId<T: Committable> {
+pub enum LeafId<Types: NodeType> {
     #[display(fmt = "{_0}")]
     Number(usize),
     #[display(fmt = "{_0}")]
-    Hash(Commitment<T>),
+    Hash(LeafHash<Types>),
 }
 
-impl<T: Committable> Clone for ResourceId<T> {
+impl<Types: NodeType> Clone for LeafId<Types> {
     fn clone(&self) -> Self {
         *self
     }
 }
 
-impl<T: Committable> PartialOrd for ResourceId<T> {
+impl<Types: NodeType> PartialOrd for LeafId<Types> {
     fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
         Some(self.cmp(other))
     }
 }
 
-pub type BlockId<Types> = ResourceId<Header<Types>>;
-pub type LeafId<Types> = ResourceId<Leaf<Types>>;
+#[derive(Derivative, From, Display)]
+#[derivative(Ord = "feature_allow_slow_enum")]
+#[derivative(
+    Copy(bound = ""),
+    Debug(bound = ""),
+    PartialEq(bound = ""),
+    Eq(bound = ""),
+    Ord(bound = ""),
+    Hash(bound = "")
+)]
+pub enum BlockId<Types: NodeType> {
```

Comment:
We have a separate type for `BlockId` now, rather than using `ResourceId`, because there are more ways to query for a block than for a leaf